draft-ietf-ppm-dap.md

```diff
@@ -942,7 +944,7 @@ struct {
   opaque agg_param<0..2^16-1>;
   select (query_type) { // determined by task configuration
     time-interval: Empty;
-    fixed-size: BatchId batch_id;
+    fixed-size: BatchID batch_id;
```

Comment:
Do we need to also make an update to `AggregateContinueReq`?

Reply:
What update are you thinking of?

On the assumption it is something along the lines of "include the batch ID in the `AggregateContinueReq` as well", I think we don't need an update to `AggregateContinueReq`: once the mapping of aggregation job to batch ID is determined by the `AggregateInitializeReq`, it is sufficient to specify only the aggregation job ID in further communications.

(apologies if I'm misinterpreting your comment!)

Reply:
Nope, you assumed right! Would the Helper need to have some sort of reverse map to recover the batch ID from the job ID?

Reply:
Thinking aloud here - 

On init, Helper gets batch and job IDs. 
On cont, Helper gets job ID.
On collect, Helper gets batchID.

So we would need batchID -> jobID and vice versa right?

Reply:
> Nope, you assumed right! Would the Helper need to have some sort of reverse map to recover the batch ID from the job ID?

Indeed: the Helper will need to track this as an additional piece of state. Presumably this reverse map would be an index (or equivalent) in a Helper implementation.

Thinking about whether this state is needed, I think it is: even if we decided that `AggregateContinueReq` should include a `BatchID` as well, I think we would want to enforce that the `BatchID` matches for all `Aggregate{Initialize,Continue}Req` messages for the same aggregation job ID, which would imply the helper storing the state there as well. (maybe there is a larger change possible?)


>Thinking aloud here -
>
>On init, Helper gets batch and job IDs.
>On cont, Helper gets job ID.
>On collect, Helper gets batchID.
>
>So we would need jobID -> batchID (for cont) and batchID -> jobID (for collect) right?

That's correct. Effectively, a batch is made up of one or more aggregation jobs. Informally, aggregation jobs are the unit of batching between the Aggregators for the majority of the aggregation process (i.e. the Aggregators evaluating the VDAF up to the point of recovering their respective output shares). Batches are the unit of collection by the Collector.

One reason that the concept of "batch" is separate from "aggregation job" is that it may be practically infeasible to group together enough reports at a time to meet the batch requirements (e.g. `min_batch_size`). Also, in many scenarios Aggregators can begin the aggregation process long before the collection process; this is aided by allowing aggregation jobs to be created even before e.g. enough reports have arrived to fill a batch. That is, the concept of aggregation jobs allows aggregating already-received reports for a batch concurrently with additional reports arriving.

Reply:
why do we need a reverse map of jobID-> batchID?

Reply:
I think the main reason we need a map from aggregation job IDs to collect batches is for enforcing collect-time privacy requirements (the aggregators will need to examine all collect batches that each aggregation job was previously a part of, given the candidate list of aggregation jobs for the current collect batch).

The reverse map may also be used during handling of AggregateContinueRequest, to look up state relating to the collect batch and incrementally incorporate output shares into an aggregate share, but that's an implementation decision. Aggregators could just as well store output shares indexed by aggregation job, and then only fetch output shares and combine them all at once when handling a CollectRequest or AggregateShareRequest.

Reply:
@branlwyd Wouldn't the simplest thing be to send both batchID and jobID on init and continue? On continue, if those two parameters don't uniquely resolve to a set of transitions, then we send back some error. On aggregate share, the Leader only sends the batchID, then the Helper finds all jobs for that batchID and can enforce the collect-time privacy requirements on the output shares for those jobs.

From an implementation perspective, you could use any distributed key value store that additionally supports a prefix query (S3 and Redis should be good candidates for this).

Ideally we can find a solution that uses a single data structure.

Reply:
IMO, your suggestion is actually more complicated, in the sense that it sends more data over the wire & creates an additional possible error case (i.e. aggregate continue req contains a mismatched aggregation job ID/batch ID) for implementations to potentially mishandle. (Resending the data does ensure that the two aggregators don't have different conceptions of the aggregation job<->batch mapping; but this would be a pretty nasty bug, and would be caught at collection time even without resending the data.)

For that reason, I'd prefer to keep the current text as written.

===

re: single data structure

I'd be amenable to a single data structure somehow merging the concept of an aggregation job & a batch; however, I don't believe the suggestion to send both batch ID & aggregation job ID moves us towards that goal. More generally, I think we'd need to somehow overcome/sidestep the practical issues I mention upthread from this comment:

> One reason that the concept of "batch" is separate from "aggregation job" is that it may be practically infeasible to group together enough reports at a time to meet the batch requirements (e.g. min_batch_size). Also, in many scenarios Aggregators can begin the aggregation process before the collection process can start; this is aided by allowing aggregation jobs to be created even before e.g. enough reports have arrived to fill a batch. That is, the concept of aggregation jobs allows aggregating already-received reports for a batch concurrently with additional reports arriving.

(To add a few more words to this: I like the ability to "unhitch" the privacy parameters, like `min_batch_size`, from the technical parameters, such as "how many client reports will we batch into a single unit of aggregation" -- I wouldn't be too happy if we ended up with a design which required trading off privacy to reach desired performance/reliability properties. And I consider the ability to commonly begin the aggregation process before the batch is full to be an important performance property of the current design.)

FWIW, I think implementations may practically already be in the world where only one data structure is required: it is possible to arrange an implementation such that "aggregation jobs" are a real/reified data structure, and "batches" exist only as a collection of aggregation jobs -- the mapping being stored with the aggregation job data structure. (In this implementation strategy, collection information is also attached to aggregation jobs, which may be surprising but also assists in ensuring that the same aggregation job is not aggregated/collected too many times.) Since in this strategy batches are represented entirely as an indexed identifier associated with an aggregation job, in this implementation strategy batches are implicit.

Reply:
Sure, the tradeoff seems to be either we send an extra field (batchID) and maintain one mapping `(batchID, jobID) -> state` on the Helper, or we only send jobID in `AggregateContinueReq` and maintain two mappings `batchID -> jobID` and `jobID -> state`. 

I was wrong earlier when I thought we would need a reverse map, so I no longer feel as strongly in either direction.

That all said, this conversation is probably out of scope for this PR and I don't want to hold this up. Happy to continue the discussion elsewhere if needed.

As for what I meant by single data structure, I was talking about the number of mappings. But I 💯 agree with the last paragraph, we were actually implementing it that way while the fixed-size proposal was iterated on. Now that Init requires the batchID to be generated, we can no longer do that I think. @branlwyd

Reply:
OK -- I'm happy to make a change if there's consensus (or if additional discussion provides even stronger justification for making this change), but will leave the text as-is for now.

===

One question (this might be moot if we are OK going with the current direction, but might influence how we iterate on this approach in the future):

> But I 100 agree with the last paragraph, we were actually implementing it that way while the fixed-size proposal was iterated on. Now that Init requires the batchID to be generated, we can no longer do that I think.

Could you elaborate on this point? I'm not sure how associating batch ID at time-of-aggregate-job-initialization blocks this implementation strategy. Also, I'm not sure what the suggested alternative is: even with the text as written at `HEAD`, i.e. before PRs #308 & #313, the text was written such aggregation jobs were associated with a batch at time of aggregation job initialization. (to my memory, #297 was always written that way--although I may be misremembering, the PR branch was squashed before merging so I've lost some of the history)

I'd like to enable implementations which represent batches "implicitly", so if I'm missing something here I'd be interested in fixing things up. (but as long as you're OK with not holding up this PR, I do think it might be valuable to continue discussion elsewhere)